### PR TITLE
PKI Redesign: setup Ember Engine

### PR DESCRIPTION
**Pull request by Monkeychip**
This PR is the first of the PKI redesign project. We will be pushing our changes to main, but keeping the work hidden behind an Ember engine (later exposed in the 1.13 release). The mount for this engine will be commented out on the `router.js` file so a user can't accidentally fall upon the new routing. **Thus, you should never see a pki mount added to the `router.js` file until the project is ready.**

image

**Notes:**
1. The new engine—similar to KMIP—can be reached when added as a mount in the router.js file as `http://localhost:4200/ui/vault/secrets/pki-1/pki/<route>` where `pki-1` is the secret engine mount's name. 

Compare this to the current route (which doesn't have the appended pki).
`http://localhost:4200/ui/vault/secrets/pki-1/<route>`

3. Ran test locally, twice in a row, including Enterprise. Everything looked good.

4. If you want to test on the ui, be sure to yarn install before hand (And you'll need to yarn install when you move off the branch too).

### CODE REVIEW

File: ui/app/templates/vault/cluster/secrets/backend/configuration.hbs [hunk 1]
```diff
@@ -6,7 +6,6 @@
     path="vault.cluster.secrets.backend.list-root"
     model=this.model.id
   }}
-  @isConfigure={{true}}
```
Comment (Monkeychip) [inline ui/app/templates/vault/cluster/secrets/backend/configuration.hbs:9]: `isConfigure` is a param not used in the SecretListHeader component. Triple checked this and reran the test suites.

File: ui/lib/core/addon/components/secret-list-header-tab.hbs [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{{#unless this.dontShowTab}}
```
Comment (Monkeychip) [inline ui/lib/core/addon/components/secret-list-header-tab.hbs:1]: Moved component to addon folder.

File: ui/lib/core/addon/components/secret-list-header.js [hunk 3]
```diff
@@ -0,0 +1,35 @@
+import Component from '@glimmer/component';
```
Comment (Monkeychip) [inline ui/lib/core/addon/components/secret-list-header.js:1]: moved to addon folder

File: ui/lib/core/addon/helpers/engines-options-for-backend.js [hunk 4]
```diff
@@ -0,0 +1,57 @@
+import { helper as buildHelper } from '@ember/component/helper';
```
Comment (Monkeychip) [inline ui/lib/core/addon/helpers/engines-options-for-backend.js:1]: duplicated the concept of this helper, but using just for PKI engine at this point. Figured in the future, we will add more so I wanted to keep this pattern. Additionally, this way I don't touch the current PKI setup in the `options-for-backend.js` helper in the non-addon helper app folder

File: ui/app/templates/vault/cluster/secrets/backend/configuration.hbs (hunk 1 above)
Comment (Monkeychip) [inline ui/app/templates/vault/cluster/secrets/backend/configuration.hbs:9]: `isConfigure` is a parma not used in the SecretListHeader. Somewhere along the way (maybe when the component was glimmerized) this param was leftover. I tripled checked and no functionality changed and the actual param was not mentioned anywhere in the `SecretListHeader` component js or hbs file.

File: ui/lib/pki/addon/engine.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+import Engine from 'ember-engines/engine';
+import loadInitializers from 'ember-load-initializers';
+import Resolver from './resolver';
+import config from './config/environment';
+
+const { modulePrefix } = config;
+/* eslint-disable ember/avoid-leaking-state-in-ember-objects */
+const Eng = Engine.extend({
```
Comment (hellobontempo) [inline ui/lib/pki/addon/engine.js:8]: Should we update this syntax to `extends Engine` - that way we can also remove the `eslint-disable` comment? 

Sample below from [ember engines quickstart](https://ember-engines.com/docs/quickstart) which is also what Chelsea's used in the Vault Insights engine [PR ](https://github.com/hashicorp/cloud-ui/pull/2936/files#diff-5ddf30e3d11c22cf19c8c1c488986e7c4363f85dec2aac48698c7a95b30acf2aR1)

```
// addon/engine.js
import Engine from '@ember/engine';

import loadInitializers from 'ember-load-initializers';
import Resolver from 'ember-resolver';

import config from './config/environment';

const { modulePrefix } = config;

export default class PkiEngine extends Engine {
  modulePrefix = modulePrefix;
  Resolver = Resolver;
}

loadInitializers(PkiEngine, modulePrefix);

File: ui/lib/pki/addon/resolver.js [hunk 6]
```diff
@@ -0,0 +1,3 @@
+import Resolver from 'ember-resolver';
```
Comment (hellobontempo) [inline ui/lib/pki/addon/resolver.js:1]: Can we delete this file if wherever we have `import Resolver from './resolver';` we change to `import Resolver from 'ember-resolver';`

File: ui/lib/pki/index.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+/* eslint-env node */
+/* eslint-disable ember/avoid-leaking-state-in-ember-objects */
+/* eslint-disable node/no-extraneous-require */
+'use strict';
+
+const EngineAddon = require('ember-engines/lib/engine-addon');
```
Comment (hellobontempo) [inline ui/lib/pki/index.js:6]: Also taken from the quickstart guide, but could we do something like this instead?

```
const { buildEngine } = require('ember-engines/lib/engine-addon');

module.exports = buildEngine({
  name: 'pki',
  lazyLoading: {
    enabled: false
  }
});

**Review by hellobontempo — COMMENTED** [review]
Just some comments! As I haven't worked much with engines I'm not confident enough in to give a full approval

**Comment by zofskeez** [comment]
I have a question about the routing. If the engine is _pki_ would it make more sense for that to come before the dyanmic engine name?
`http://localhost:4200/ui/vault/secrets/pki/<engine-name>/<route>`

File: ui/lib/pki/addon/engine.js (hunk 5 above)
Comment (Monkeychip) [reply ui/lib/pki/addon/engine.js:8]: Thank you for pointing this out!  I'll update.

File: ui/app/router.js [hunk 8]
```diff
@@ -114,6 +114,7 @@ Router.map(function () {
         this.route('backends', { path: '/' });
         this.route('backend', { path: '/:backend' }, function () {
           this.mount('kmip');
+          // this.mount('pki'); // ARG TODO always comment out when pushing to main
```
[the remark this answers is not in the document]
Comment (Monkeychip) [reply ui/app/router.js]: Great idea! Amended, and tested.

**Review by zofskeez — APPROVED** [review]
Looks like you're off to a good start. Thanks!

**Pull request merged**